lib/spack/spack/directives.py

```diff
@@ -863,6 +864,29 @@ def _execute_maintainer(pkg):
     return _execute_maintainer
 
 
+@directive("licenses")
+def license(license_identifier: str, when=None):
+    """Add a new license directive, to specify the SPDX identifier the software is
+    distributed under.
+
+    Args:
+        license_identifiers: A list of SPDX identifiers specifying the licenses
+            the software is distributed under.
+        when: A spec specifying when the license applies.
+    """
+
+    def _execute_license(pkg):
+        # If when is not specified the conflict always holds
+        when_spec = make_when_spec(when)
+        if not when_spec:
+            return
+
+        when_spec_list = pkg.licenses.setdefault(license_identifier, [])
+        when_spec_list.append(when_spec)
```

Comment:
I'm in the middle of a refactor on how we store data on packages.  Short story: all dictionaries on the package are going to be keyed by `when` spec.  So, can you invert this one?  Basically instead of storing a dict like:

```
licenses = { license_id: [when_spec, ...] }
```

Do this instead:

```
licenses = { when_spec: license_id }
```

I think the directive should also raise an error if you try to specify different license ids for the same `when` spec, or, really, if any of the `when` specs overlap.  You can easily check the former as the directive is called (basically if there is already a license id and it's different, complain).  The latter probably needs to be done after all the meta class stuff executes... you can use `Spec.intersects()` to do it.  e.g. `foo+bar` and `foo~bar` do not intersect but `foo` and `foo+bar` do (b/c there are concrete specs that could match both).

Reply:
I've implemented checking for overlapping specs by just calling `Spec.intersect` within the `_execute_license` function. It seems to cover the cases that I can think of. I guess that's only run as the meta class is executing though?